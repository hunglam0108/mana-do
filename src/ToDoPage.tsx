import React from 'react';
import {useEffect, useReducer, useRef, useState} from 'react';
import { Button, Input, Space, Form } from 'antd';
import { CloseOutlined } from '@ant-design/icons';
import reducer, {initialState} from './store/reducer';
import {
    setTodos,
    createTodo,
    toggleAllTodos,
    deleteAllTodos,
    updateTodoStatus,
    activeAllTodos,
    completeAllTodos,
    deleteTodo,
    editTodo
} from './store/actions';
import Service from './service';
import {TodoStatus} from './models/todo';

type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = () => {
    const [{todos}, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
    const inputRef = useRef<any>(null);
    const inputEditRef = useRef<any>(null);
    const [edit, setEdit] = useState(false);
    const [keyEdit, setKeyEdit] = useState(0);
    const wrapperRef = useRef<HTMLDivElement>(null);
    const [form] = Form.useForm();
    useEffect(()=>{
        (async ()=>{
            const resp = await Service.getTodos();
            dispatch(setTodos(resp || []));
            document.addEventListener('click', handleClickOutside, true);
        })()
    }, [])

    const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === 'Enter' && form.getFieldValue("input-create") ) {
            const resp = await Service.createTodo(form.getFieldValue("input-create"));
            dispatch(createTodo(resp));
        }
    }

    const onEditTodo = async (e: React.KeyboardEvent<HTMLInputElement>, todoId: string) => {
        if (e.key === 'Enter' && form.getFieldValue("edit-create") ) {
            dispatch(editTodo(todoId, form.getFieldValue("edit-create")));
            setEdit(false)
        }
    }

    const onUpdateTodoStatus = (e: any, todoId: string) => {
        console.log(e.target.checked)
        dispatch(updateTodoStatus(todoId, e.target.checked))
    }

    const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(toggleAllTodos(e.target.checked))
    }

    const onActiveAllTodo = () => {
        dispatch(activeAllTodos())
    }

    const onCompleteAllTodo = () => {
        dispatch(completeAllTodos())
    }

    const onDeleteAllTodo = () => {
        dispatch(deleteAllTodos());      
    }

    const onDeleteTodo = (todoId: string) => {
        dispatch(deleteTodo(todoId));        
    }

    const handleClickOutside = (event: Event) => {
        if (
            wrapperRef.current &&
            !wrapperRef.current.contains(event.target as Node)
        ) {
            setEdit(false);
        }
    };
    return (
        <Space className="ToDo__container" direction="vertical" size="middle">
            <Form form={form}>
            <Form.Item className="Todo__creation" name="input-create"
             rules={[
                { required: true, message: "Not input blank value !!!" }]}
            >
                <Input
                    ref={inputRef}
                    className="Todo__input"
                    data-testid= "Todo__input"
                    placeholder="What need to be done?"
                    onKeyDown={onCreateTodo}
                />
            </Form.Item>
            <div className="ToDo__list">
                {
                    todos.map((todo, index) => {
                        return (
                            //index is number not id, so onUpdateTodoStatus must have para todo.id
                            <div key={index} className="ToDo__item" data-testid= "Todo__item">
                                <input
                                    type={"checkbox"}
                                    checked={ todo.status === TodoStatus.COMPLETED }
                                    onChange={(e) => { onUpdateTodoStatus(e, todo.id) }}
                                />
                                {/* <span>{todo.content}</span> */}
                                {(edit && keyEdit === index) ? 
                                    (<div ref={wrapperRef}> 
                                        <Form.Item name="edit-create" className="ToDo__edit" initialValue={todo.content}
                                        rules={[
                                            { required: true, message: "Not input blank value !!!" }]}
                                        >
                                        <Input ref={inputEditRef} style={{width: 350}} type='text' onKeyDown={ (e) => onEditTodo(e, todo.id) } />
                                        </Form.Item>
                                    </div>) : 
                                    (<span onDoubleClick={() => { setEdit(true); setKeyEdit(index) }}>{todo.content}</span>)}
                                                             
                                <Button
                                    className="Todo__delete"
                                    onClick={() => {onDeleteTodo(todo.id)}}
                                    icon={<CloseOutlined/>}
                                >
                                </Button>
                            </div>
                        );
                    })
                }
            </div>
            </Form>
            <div className="Todo__toolbar">
                {todos.length > 0 ?
                    <input
                        type="checkbox"
                        onChange={onToggleAllTodo}
                    /> : <div/>
                }
                <div className="Todo__tabs">
                    <Button type={"primary"}>
                        All
                    </Button>
                    <Button type={"primary"} onClick={()=> onActiveAllTodo()}>
                        Active
                    </Button>
                    <Button type={"primary"} onClick={()=> onCompleteAllTodo() }>
                        Completed
                    </Button>
                </div>
                <Button type={"primary"} onClick={onDeleteAllTodo}>
                    Clear all todos
                </Button>
            </div>
        </Space>
    );
};

export default ToDoPage;