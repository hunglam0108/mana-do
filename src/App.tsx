import React from 'react';

import ToDoPage from './ToDoPage';
import "antd/dist/antd.css";
import './App.css';

function App() {
  return (
    <main className="App">
      <ToDoPage />
    </main>
  );
}

export default App;
