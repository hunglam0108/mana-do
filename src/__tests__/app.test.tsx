import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ToDoPage from '../ToDoPage';
import enzyme from 'enzyme';
import { render, fireEvent, getByTestId} from "@testing-library/react";
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });
describe("Test", () => {
    beforeAll(() => {
      Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation((query: any) => ({
          matches: false,
          media: query,
          onchange: null,
          addListener: jest.fn(), // Deprecated
          removeListener: jest.fn(), // Deprecated
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
          dispatchEvent: jest.fn(),
        }))
      });
    });
    
    test('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<App />, div);
      });
    test('test ToDoPage', () => {
        const component = renderer.create(<ToDoPage />).toJSON();
        expect(component).toMatchSnapshot();
    });
    test('test event', () => {
        const { container, rerender } = render(<App />);
        const input = getByTestId(container, "Todo__input");
        fireEvent.change(input, {target: {value: 'todo 1'}})
        fireEvent.keyDown(input, { key: "Enter", code: 13 })
        const item = getByTestId(container, "Todo__item");
        expect(item).toBeTruthy()
    });
});
