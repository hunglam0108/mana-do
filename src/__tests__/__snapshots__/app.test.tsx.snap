// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test test ToDoPage 1`] = `
<div
  className="ant-space ant-space-vertical ToDo__container"
  style={Object {}}
>
  <div
    className="ant-space-item"
    style={
      Object {
        "marginBottom": 16,
      }
    }
  >
    <form
      className="ant-form ant-form-horizontal"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="ant-row ant-form-item Todo__creation"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <input
                className="ant-input Todo__input"
                data-testid="Todo__input"
                id="input-create"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="What need to be done?"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ToDo__list"
      />
    </form>
  </div>
  <div
    className="ant-space-item"
    style={Object {}}
  >
    <div
      className="Todo__toolbar"
    >
      <div />
      <div
        className="Todo__tabs"
      >
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            All
          </span>
        </button>
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            Active
          </span>
        </button>
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            Completed
          </span>
        </button>
      </div>
      <button
        className="ant-btn ant-btn-primary"
        onClick={[Function]}
        type="button"
      >
        <span>
          Clear all todos
        </span>
      </button>
    </div>
  </div>
</div>
`;
